### ISSUE TITLE
Unable to make aggregation query

### ISSUE THREAD

**Issue by NielsKSchjoedt** [googleapis/google-cloud-ruby#19951]
I am trying to run an aggregation query using [GoogleCloudDatastore SDK](https://github.com/googleapis/google-cloud-ruby/tree/main/google-cloud-datastore) (a simple `count` of matching entities). However the API does not seem to support it, as it throws an error:

```ruby
data_store = ::Google::Cloud::Datastore.new
q = data_store.gql("SELECT count(*) FROM `MyKind` WHERE source_name = 'MySourceName'")
data_store.run(q)
```

resulting in:

```
/usr/local/bundle/gems/google-cloud-datastore-v1-0.7.0/lib/google/cloud/datastore/v1/datastore/client.rb:370:in `rescue in run_query': 3:Expecting ENTITY_QUERY, but got AGGREGATION_QUERY. Use Datastore.RunAggregationQuery API instead for such queries.. debug_error_string:{UNKNOWN:Error received from peer ipv4:142.250.184.234:443 {created_time:"2023-01-04T13:54:56.689301852+00:00", grpc_status:3, grpc_message:"Expecting ENTITY_QUERY, but got AGGREGATION_QUERY. Use Datastore.RunAggregationQuery API instead for such queries."}} (Google::Cloud::InvalidArgumentError)
```

I am able to do it from the cloud web console:

![datastore cloud console](https://user-images.githubusercontent.com/570755/210742706-f8e0ab3c-6c32-4249-8db3-b14e2f4401b6.jpg)

I also tried using [GoogleCloudDatstoreV1 SDK](https://github.com/googleapis/google-cloud-ruby/tree/main/google-cloud-datastore-v1) and this [snippet](https://github.com/googleapis/google-cloud-ruby/blob/main/google-cloud-datastore-v1/snippets/datastore/run_aggregation_query.rb) :

```ruby
request = Google::Cloud::Datastore::V1::RunAggregationQueryRequest.new(aggregation_query: "SELECT count(*) FROM MyKind WHERE source_name = 'MySourceName'")
```

But it throws:

```
Google::Protobuf::TypeError: Invalid type String to assign to submessage field 'aggregation_query'. f
rom (pry):3:in initialize'
```

same for:

```ruby
request = Google::Cloud::Datastore::V1::RunAggregationQueryRequest.new(gql_query: "SELECT count(*) FROM MyKind WHERE source_name = 'MySourceName'")
```

```
Google::Protobuf::TypeError: Invalid type String to assign to submessage field 'gql_query'. 
from (pry):4:in initialize'
```

I'm clueless of this...?

**Comment by NielsKSchjoedt** [comment]
@diptanshumittal I think I found something, however now there seems to be segfault:

```ruby
client = Google::Cloud::Datastore::V1::Datastore::Client.new

q = ::Google::Cloud::Datastore::V1::GqlQuery.new
q.allow_literals=true
q.query_string="SELECT count(*) FROM MyKind WHERE source_name = 'MySourceName'"

request = Google::Cloud::Datastore::V1::RunAggregationQueryRequest.new(gql_query: q)

client.run_aggregation_query(request)
```

results in ruby crashing:

```
objc[97515]: +[__NSTimeZone initialize] may have been in progress in another thread when fork() was called.
objc[97515]: +[__NSTimeZone initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

```
ruby -v #=> ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin21]
```

**Comment by diptanshumittal** [comment]
> I am trying to run an aggregation query using [GoogleCloudDatastore SDK](https://github.com/googleapis/google-cloud-ruby/tree/main/google-cloud-datastore) (a simple `count` of matching entities). However the API does not seem to support it, as it throws an error:
> 
> ```ruby
> data_store = ::Google::Cloud::Datastore.new
> q = data_store.gql("SELECT count(*) FROM `MyKind` WHERE source_name = 'MySourceName'")
> data_store.run(q)
> ```
> 
> resulting in:
> 
> ```
> /usr/local/bundle/gems/google-cloud-datastore-v1-0.7.0/lib/google/cloud/datastore/v1/datastore/client.rb:370:in `rescue in run_query': 3:Expecting ENTITY_QUERY, but got AGGREGATION_QUERY. Use Datastore.RunAggregationQuery API instead for such queries.. debug_error_string:{UNKNOWN:Error received from peer ipv4:142.250.184.234:443 {created_time:"2023-01-04T13:54:56.689301852+00:00", grpc_status:3, grpc_message:"Expecting ENTITY_QUERY, but got AGGREGATION_QUERY. Use Datastore.RunAggregationQuery API instead for such queries."}} (Google::Cloud::InvalidArgumentError)
> ```
> 
> I am able to do it from the cloud web console:
> 
> ![datastore cloud console](https://user-images.githubusercontent.com/570755/210742706-f8e0ab3c-6c32-4249-8db3-b14e2f4401b6.jpg)
> 
> I also tried using [GoogleCloudDatstoreV1 SDK](https://github.com/googleapis/google-cloud-ruby/tree/main/google-cloud-datastore-v1) and this [snippet](https://github.com/googleapis/google-cloud-ruby/blob/main/google-cloud-datastore-v1/snippets/datastore/run_aggregation_query.rb) :
> 
> ```ruby
> request = Google::Cloud::Datastore::V1::RunAggregationQueryRequest.new(aggregation_query: "SELECT count(*) FROM MyKind WHERE source_name = 'MySourceName'")
> ```
> 
> But it throws:
> 
> ```
> Google::Protobuf::TypeError: Invalid type String to assign to submessage field 'aggregation_query'. f
> rom (pry):3:in initialize'
> ```
> 
> same for:
> 
> ```ruby
> request = Google::Cloud::Datastore::V1::RunAggregationQueryRequest.new(gql_query: "SELECT count(*) FROM MyKind WHERE source_name = 'MySourceName'")
> ```
> 
> ```
> Google::Protobuf::TypeError: Invalid type String to assign to submessage field 'gql_query'. 
> from (pry):4:in initialize'
> ```
> 
> I'm clueless of this...?

Value for aggregation_query and gql_query is invalid as it doesn't accept string.

**Comment by diptanshumittal** [comment]
> @diptanshumittal I think I found something, however now there seems to be segfault:
> 
> ```ruby
> client = Google::Cloud::Datastore::V1::Datastore::Client.new
> 
> q = ::Google::Cloud::Datastore::V1::GqlQuery.new
> q.allow_literals=true
> q.query_string="SELECT count(*) FROM MyKind WHERE source_name = 'MySourceName'"
> 
> request = Google::Cloud::Datastore::V1::RunAggregationQueryRequest.new(gql_query: q)
> 
> client.run_aggregation_query(request)
> ```
> 
> results in ruby crashing:
> 
> ```
> objc[97515]: +[__NSTimeZone initialize] may have been in progress in another thread when fork() was called.
> objc[97515]: +[__NSTimeZone initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
> ```
> 
> ```
> ruby -v #=> ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin21]
> ```

Error not reproducable.. Can you please help by providing the entire stack?

**Comment by SandeepTuniki** [comment]
@NielsKSchjoedt The aggregation feature will soon be available in Datastore. You can keep a track of it's progress in #20039.

In general, we don't recommend using the `Datastore::V1` library directly, since it requires deeper understanding of the protos & grpc. So instead rely on the wrapper `Datastore` library (via `require 'google/cloud/datastore'`).

**Comment by SandeepTuniki** [comment]
@NielsKSchjoedt This feature is now available in Datastore (released in version `2.4.0` in #20110). 

It's example usage is:

```rb
require "google/cloud/datastore"

datastore = Google::Cloud::Datastore.new

# Query on which to aggregate
query = Google::Cloud::Datastore::Query.new
query.kind("Task")
     .where("done", "=", false)

# Construct an aggregate query
aggregate_query = query.aggregate_query
                       .add_count

# Query the backend
aggregate_query_results = dataset.run_aggregation aggregate_query
puts aggregate_query_results.get

```

I'm closing this issue, but feel free to re-open it if you face any further issue.
